onmt/translate/translator.py

```diff
@@ -622,15 +524,23 @@ def _translate_batch(
             src_vocabs,
             max_length,
             min_length=0,
-            ratio=0.,
-            n_best=1,
+            beam_size=1,
```

Comment:
`_translate_batch` should probably not know about the beam size. I suggest building the decoding strategy in `translate_batch` and renaming `_translate_batch` to `_translate_batch_with_strategy`.

Reply:
I was intending to do so, but later I realize that we need `memory_length` and `device` to instantialize the decoding strategy. These two arguments are created after we've done the run_encoder.
To build the decoding strategy in `translate_batch`, we need to separate current DecodeStrategy(BeamSearch/RandomSampling).\_\_init\_\_ in 2 step, the part not concern `device` or `memory_length` as new \_\_init\_\_, and those need `device` or `memory_length` as a initialize method to be called at (step "(2)") and return tiled `src_map` and `memory_bank`.
Is this what you mean?

Reply:
Yes. Arguments that are dependent on the runtime (e.g. device, encoder output, etc.) should be passed to an `initialize` method and not to the constructor.